apps/server/lib/lexical/server/code_intelligence/entity.ex

```diff
@@ -0,0 +1,211 @@
+  alias Future.Code, as: Code
+  alias Lexical.Document
+  alias Lexical.Document.Location
+  alias Lexical.Document.Position
+  alias Lexical.Document.Range
+  alias Lexical.Project
+  alias Lexical.RemoteControl
+  alias Lexical.RemoteControl.CodeIntelligence
+  alias Lexical.RemoteControl.CodeMod
+  alias Lexical.Text
+  alias Sourceror.Zipper
+
+  require Logger
+
+  @type resolved :: {:module, module()}
+
+  @doc """
+  Attempts to resolve the entity at the given position in the document.
+
+  ## Return values
+
+  Returns `{:ok, resolved}` if successful and `{:error, error}` otherwise.
+  Resolved entities are one of:
+
+    * `{:module, module}`
+
+  """
+  @spec resolve(Document.t(), Position.t()) :: {:ok, resolved} | {:error, term()}
+  def resolve(%Document{} = document, %Position{} = position) do
+    with {:ok, ast} <- CodeMod.Ast.from(document),
+         {:ok, zipper} <- innermost_zipper_at(ast, position),
+         {:ok, resolved} <- resolve(Zipper.node(zipper), zipper, document, position) do
+      Logger.info("Resolved entity: #{inspect(resolved)}")
+      {:ok, resolved}
+    else
+      error ->
+        Logger.info("Resolve failed: #{inspect(error)}")
+        error
+    end
+  end
+
+  defp resolve({:__aliases__, _, aliases} = node, zipper, document, position) do
+    with {:ok, %{start: start}} <- fetch_range(node) do
+      {aliases_with_columns, _} =
+        Enum.map_reduce(aliases, start[:column], fn
+          {:__MODULE__, _, nil}, column ->
+            {{:__MODULE__, column}, column + 10 + 1}
+
+          alias_name, column ->
+            alias_len = alias_name |> Atom.to_string() |> String.length()
+            {{alias_name, column}, column + alias_len + 1}
+        end)
+
+      at_or_before_position =
+        for {name, offset} <- aliases_with_columns,
+            offset <= position.character do
+          name
+        end
+
+      case at_or_before_position do
+        [:__MODULE__ | rest] ->
+          with {:ok, module_aliases} <- current_module_aliases(zipper) do
+            {:ok, {:module, Module.concat(module_aliases ++ rest)}}
+          end
+
+        aliases ->
+          {:ok, module} = CodeIntelligence.Ast.expand_aliases(document, position, aliases)
+          {:ok, {:module, module}}
+      end
+    end
+  end
+
+  defp resolve({:__MODULE__, _, nil}, zipper, _document, _position) do
+    with {:ok, module_aliases} <- current_module_aliases(zipper) do
+      {:ok, {:module, Module.concat(module_aliases)}}
+    end
+  end
+
+  defp resolve(node, _zipper, _document, _position) do
+    {:error, {:unsupported, node}}
+  end
+
+  defp current_module_aliases(nil), do: {:error, :missing_defmodule}
+
+  defp current_module_aliases(zipper) do
+    case Zipper.node(zipper) do
+      {:defmodule, _, [{:__aliases__, _, aliases} | _]} -> {:ok, aliases}
+      _ -> zipper |> Zipper.up() |> current_module_aliases()
+    end
+  end
+
+  defp innermost_zipper_at(ast, position) do
```

Comment:
I spent some time trying to refactor this to get past the Credo check, but couldn't come up with anything I liked better. Suggestions welcome.

Reply:
pretty much the only way to fix cyclomatic complexity errors is to pull out functions. I actually think credo calculates CC incorrectly. This doesn't feel crazy to me, but the nested function + case + cond is likely what's bumping up the CC.